sysinit/coqargs.ml

```diff
@@ -448,7 +448,12 @@ let build_load_path opts =
     else
       let coqenv = Boot.Env.init () in
       Coqloadpath.init_load_path ~coqenv in
-  ml_path @ opts.pre.ml_includes ,
+  let ocamlpathsep = if Sys.unix then ':' else ';' in
+  let env_ocamlpath =
+    try String.split_on_char ocamlpathsep (Sys.getenv "OCAMLPATH")
+    with Not_found -> []
+  in
+  opts.pre.ml_includes @ env_ocamlpath @ ml_path ,
```

Comment:
Is it on purpose that the content of `$OCAMLPATH` is passed to `Mltop.add_ml_dir`? This seems like a very bad idea.

Reply:
This point came up at the call, but nobody could find the problem. What is it?

Reply:
What problem?

Reply:
> the content of $OCAMLPATH is passed to Mltop.add_ml_dir? This seems like a very bad idea.

what is the problem when doing that?

Reply:
I guess not a critical one.

I'd be nice tho if we can find some cycles in Sep and remove the legacy mode, actually there is not a lot to do, but what requires time (as you are well aware) is testing.

Reply:
The problem is that, on some systems (e.g., Nix), `$OCAMLPATH` contains the whole world. Thus, when trying to load a legacy plugin, Coq will scan a bunch of directories that are completely unrelated to Coq. (Notice that this pull request scans `env_ocamlpath` before `ml_path`, where the legacy plugins lie.) In fact, if by sheer luck Coq were to actually find a plugin with the searched name in one of these directories, it is guaranteed to be a broken one.

Reply:
OK, then I think @SkySkimmer should avoid this, and that requires changing the API IMO, since you need to separate the paths from OCAMPATH from the others.

I keep the occasion, in case Nix people are listening, to remind that OCAMLPATH is there to let one override the system config, not a place to store it. If they want to stay with env variables, they should add a nix specific one to findlib, eg NIX_OCAMLPATH, and have findlib honor both, giving OCAMLPATH precedence.

Reply:
You only need to change the API if you want `$OCAMLPATH` to be searched after `opts.pre.ml_includes` but before `ml_path`. But why do you even want that in the first place? Why should the Coq directories in `ml_path` not take precedence over `OCAMLPATH`?

Reply:
> remind that OCAMLPATH is there to let one override

because of this?

Reply:
> I keep the occasion, in case Nix people are listening, to remind that OCAMLPATH is there to let one override the system config, not a place to store it. If they want to stay with env variables, they should add a nix specific one to findlib, eg NIX_OCAMLPATH, and have findlib honor both, giving OCAMLPATH precedence.

Nix is using this "trick" in many programming languages that have a `PATH` environment variable, and has done so successfully for years. If Nix users had had to request changes to most of their programming tools to be able to work, it would have been less successful. The fact that Nix is becoming more and more important and has more and more users does not justify (IMHO) to introduce Nix-specific quirks in generic tools.